### PR TITLE
[ENH] Implement simple(r) benchmarking

### PR DESCRIPTION
**Pull request by DBCerigo**
#### Reference Issues/PRs
https://github.com/aeon-toolkit/aeon/issues/116 for general context. 
https://github.com/aeon-toolkit/aeon/issues/116#issuecomment-1509958967 for context specific to this PR and implementation.  

#### What does this implement/fix? Explain your changes.

- a core benchmarking module
- forecasting benchmarking module (using the core module)
- unit tests
- example notebook showing usage

#### Does your contribution introduce a new dependency? If yes, which one?

Optional, to `all_extras`: `kotsu>=0.3.1`

#### What should a reviewer concentrate their feedback on?

Reviewing the example notebook first to get a sense of usage of the API and if it fits with how you envisaged it.
Then the implementation.  

Considerations for further work: 
- Deleting/archiving all the previous benchmarking code?
- Implementing benchmarking modules for classification, clustering, etc.

#### Did you add any tests for the change?

Yes.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

### CODE REVIEW

File: aeon/benchmarking/benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        """Run the benchmark.
+
+        Parameters
+        ----------
+        output_file : str
+            Path to write results output file to.
+
+        Returns
+        -------
+        pandas DataFrame of results
```
Comment (hadifawaz1999) [inline aeon/benchmarking/benchmarks.py:73]: I think it would be better to show the output format of the Data Frame here.

Comment (hadifawaz1999) [reply aeon/benchmarking/benchmarks.py:73]: For instance the column names, rows etc. one line sentence

**Comment by TonyBagnall** [comment]
one thing is that you need a callable dataset_loader. For machine learning we have a single function (load_ts_file) we pass the location of the file to load as a string. This implies you need a different function for each datasets, not sure how that would work?

**Comment by ltsaprounis** [comment]
> one thing is that you need a callable dataset_loader. For machine learning we have a single function (load_ts_file) we pass a string of the location of the file to load. This set up implies you need a different function for each datasets, not sure how that would work?

Could always do something like this to make it a callable.

```python
# Option 1
from functools import partial

load_data = partial(load_ts_file, path="path")

# Option 2
def load_data():
    return load_ts_file("path")
```

But I guess it's more of a headache...

@DBCerigo what's the idea behind having the loader as a callable and not just passing the data?

File: aeon/benchmarking/benchmarks.py (hunk 1 above)
Comment (DBCerigo) [reply aeon/benchmarking/benchmarks.py]: Done [e3f2bde](https://github.com/aeon-toolkit/aeon/pull/379/commits/e3f2bde24ecede2cf081c8641fd09283063c2063)

**Comment by DBCerigo** [comment]
> one thing is that you need a callable dataset_loader. For machine learning we have a single function (load_ts_file) we pass a string of the location of the file to load. This set up implies you need a different function for each datasets, not sure how that would work?

@TonyBagnall @ltsaprounis 

As shown in the example notebook https://github.com/datavaluepeople/aeon/blob/benchmarking/v2/examples/04_benchmarking_v2.ipynb the intention with using `dataset_loader` is to match the module `aeon.datasets` (for example `from aeon.datasets import load_airline` is used in the notebook) in keeping with interfacing directly with `aeon` objects and classes. The notebook also shows a method for adding your own datasets - the same as Option 2 in the above. 

If `aeon` is moving away from using the `aeon.datasets` module, then I suggest we alter the implementation to not be so focussed on interfacing well with that module. If instead we wanted to only support usage via `load_ts_file` we could make `add_task` take a `dataset_path: str` (instead of `dataset_loader: Callable`) and alter this line
 https://github.com/datavaluepeople/aeon/blob/e3f2bde24ecede2cf081c8641fd09283063c2063/aeon/benchmarking/forecasting.py#L37
 to use `load_ts_file`. 

Let me know yours thought/preferences.

**Comment by hadifawaz1999** [comment]
How about taking the dataset as parameter ? Because not all users work with ts files.

**Comment by ltsaprounis** [comment]
I don't mind it being callable. It allows for some flexibility down the line e.g. if `load_ts_file` returns metadata too etc. For forecasting benchmarking I think this is good to go, there's going to be some more work to be done to allow for hierarchical data benchmarking etc. but for now I think this gives as a good starting point to start working on this.

@DBCerigo - that being said, I suggest we add a warning the public methods / classes, saying that this interface is experimental and we might change the API in the future.
```python
warn("<name of class/function> is experimental, and interfaces may change.")
```

**Comment by DBCerigo** [comment]
> How about taking the dataset as parameter ? Because not all users work with ts files.

I don't think we would want to have the datasets in memory at the time we are add benchmarking tasks, given we might be adding tasks for many multiple different datasets to compare models over etc. 

@ltsaprounis all sounds good to me. I'll wait for a day or two to see if anyone else pitches in further, and if not then I'll add the warnings and leave it to be merged if so.

**Comment by DBCerigo** [comment]
@ltsaprounis now done. (And fixed missing [ENH] tag on previous commit message.)

**Comment by TonyBagnall** [comment]
I'm happy for this to go in, but will clarify our use case. The datasets module is staying, but it only contains toy examples. Most of the problems we use are on timeseriesclassification.com, and we load them from disk with

def load_from_tsfile(
    full_file_path_and_name,
    replace_missing_vals_with="NaN",
    return_meta_data=False,
)
I'm sure we can work something out for classification, if this is ok for forecasting then go for it.

**Comment by DBCerigo** [comment]
> I'm happy for this to go in, but will clarify our use case. The datasets module is staying, but it only contains toy examples. Most of the problems we use are on timeseriesclassification.com, and we load them from disk with

I'm confident we could (easily) extend the interface in the future to be able to take just the path to a file on disk and load it (with `load_from_tsfile`) at validation run time (while still keeping the funcs which return data option as well), if we find that that functionality is much desired. 

> I'm sure we can work something out for classification, if this is ok for forecasting then go for it.

The intention with the implementation and this PR, is that `aeon/benchmarking/forecasting.py` serves as a template for implementing benchmarking for classification (and clustering, etc.). It won't be quite as simple as replacing `forecasting` with `classification` in the file, but it shouldn't be too too far off :)

**Review by ltsaprounis — APPROVED** [review]
good stuff 🚀

**Pull request merged**